### PR TITLE
smb: New keyword smb.cmd v3

### PR DESCRIPTION
**Pull request by zer1t0**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: 
https://redmine.openinfosecfoundation.org/issues/5069

Describe changes:
- New keyword smb.cmd added, which allows to match the command value in smb message header. It works for smb1 and smb2.
- Documentation for the keyword added (created SMB keyword documentation)

Example of rule `alert smb any any -> any any (msg: "Smb command rule"; smb.cmd: 10; sid: 1;)` . More examples in documentation.

suricata-verify-pr: 733

### CODE REVIEW

**Review by catenacyber — COMMENTED** [review]
Looks fine enough for me, even if there are still some nits that can be improved

- CI : ✅ 
- Code : looks good to me
- Commits segmentation : ok as 1
- Commit messages : Could you add the redmine ticket number in the description ?
- Git ID set : looks fine for me
- CLA : I do not have access, but looks like some commits were already merged
- Doc update : ok, could you add an example with a command as it is logged ?
- Redmine ticket : ok
- Rustfmt : was not enforced
- Tests : Suricata-verify tests look fine

File: rust/src/smb/detect.rs [hunk 1]
```diff
@@ -201,3 +206,433 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
...
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_free(ptr: *mut c_void) {
+    if ptr != std::ptr::null_mut() {
+        std::mem::drop(Box::from_raw(ptr as *mut SmbCmdData));
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct SmbCmdData (HashMap<u8, HashSet<u16>>);
```
Comment (catenacyber) [inline rust/src/smb/detect.rs:262]: Could you please comment what is this structure ? With regards to SMB1 and SMB2 ?

File: src/detect-smb-cmd.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/* Copyright (C) 2017 Open Information Security Foundation
```
Comment (catenacyber) [inline src/detect-smb-cmd.h:1]: nit s/2017/2022/

File: src/detect-smb-cmd.c [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    SCLogDebug("smb_cmd: DetectSmbCmdMatchRust");
+
+    if (rs_smb_cmd_match(txv, (void *)m) != 1)
+        SCReturnInt(0);
+
+    SCReturnInt(1);
+}
+
+static int DetectEngineInspectSmbCmd(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
```
[the remark this answers is not in the document]
Comment (zer1t0) [reply src/detect-smb-cmd.c:92]: Tests fail if that part is removed.

**Comment by zer1t0** [comment]
Changes applied and moved to https://github.com/OISF/suricata/pull/7181

File: rust/src/smb/smb1.rs [hunk 4]
```diff
@@ -53,69 +53,138 @@ pub const SMB1_COMMAND_PROCESS_EXIT:            u8 = 0x11;
 pub const SMB1_COMMAND_SEEK:                    u8 = 0x12;
 pub const SMB1_COMMAND_LOCK_AND_READ:           u8 = 0x13;
 pub const SMB1_COMMAND_WRITE_AND_UNLOCK:        u8 = 0x14;
+pub const SMB1_COMMAND_READ_RAW:                u8 = 0x1a;
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/smb/smb1.rs:56]: would also love to see a SV test with this record type

File: src/detect-smb-cmd.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-smb-cmd.c:92]: This is strange `http2.frametype` does not need this and I do not see the difference...

**Pull request closed without merging**